### DISCUSSION TITLE
ERROR    'utf-8' codec can't decode byte 0x85 in position 0: invalid start byte

### DISCUSSION THREAD

**Discussion by bnvader** [oracle/python-oracledb#335] [General]
I am trying to use python oracledb to connect to Oracle Database that has encoding - ISO-8859-1.
I have a custom database type of type DB Table.
When I use thick client mode, I am able to directly use the DBObject.Attribute_Name to retrieve the attribute values from a row.
However this same step fails if I do not use thick mode.

I would like to have this working in thin mode as I want to eventually use this from a simple AWS Lambda function.

Has anyone been able to use thin mode to successfully decode non utf-8 encoded Oracle DB?
If so what is the secret?

**Comment by anthony-tuininga** [comment]
The first step is to confirm what the database character set actually is: https://python-oracledb.readthedocs.io/en/latest/user_guide/globalization.html#database-character-set. Once you confirm this, can you supply it here along with the DDL required to create the object type and the Python code you are using to access it?

If the data stored in the database actually follows this character set, then conversion to AL32UTF8 (utf-8) is automatic and will not result in errors. It is, however, quite possible for the client and server to specify the SAME character set and then the server doesn't perform any conversions or checks. In that case the client can store corrupt data without any errors being raised! If that is the case you can use this technique to work around it: https://python-oracledb.readthedocs.io/en/latest/user_guide/sql_execution.html#querying-corrupt-data.

**Reply by bnvader** [reply]
Here are the results from db: 
```
NLS_CHARACTERSET' - AL32UTF8
NLS_NCHAR_CHARACTERSET - AL16UTF16
```

The ddl of the custom types - 
```
create or replace TYPE "TRANSACTION_XML_OBJ"                                          
as object
 (trans_id number
 ,status varchar2(16)
 ,last_modified_dt date 
 ,user_data xmltype
 ,channel varchar2(256)
 )
create or replace TYPE  "TRANSACTION_XML_OBJ_LIST"                                          
as table of Transaction_XML_OBJ
```

We have a stored procedure that returns the above list.
When iterate over the objects in the result list and try to do result_obj.channel, it is throwing the decoding error in thin mode.
Thick mode works fine without having to specify any encoding.
It shows the correct type for the DbObject from the list.

```
p_trans_list_type = connection.gettype("TRANSACTION_XML_OBJ_LIST")
p_list= p_trans_list_type.newobject()
```

... invoke the stored proc...

```
for resultobj in p_list.aslist():
    channel = str(resultobj.CHANNEL)
```

**Reply by anthony-tuininga** [reply]
Where does the encoding ` ISO-8859-1` come into play? The database uses AL32UTF8 (aka utf-8) so it should transfer the data the same way in both thick and thin modes. Can you supply the full traceback? It would also be useful to show the packet data in thin mode by setting the environment variable `PYO_DEBUG_PACKETS` to any value. Only the packets that show the procedure call and response from the database are required. Please also include the usual information: Python version, python-oracledb version, Oracle Database version, Oracle Client version (for thick mode).

**Reply by bnvader** [reply]
Perhaps we can ignore the ISO encoding part for now. It is used to encrypt some sensitive data in one of the columns. However it is base64 encoded eventually, so probably that is not an issue here.
I will try to get the trace packets. Might take some time as we are currently exploring enabling the thick mode in AWS.
If that works, we will just use thick mode.
Versions- 
Oracle DB - 19c on AWS RDS
Python - 3.12
Oracledb - 2.1.2
Oracledb Windows Client - thick_impl.cp312-win_amd64.pyd

**Reply by anthony-tuininga** [reply]
Sure. It would be nice to get to the bottom of this but I understand if you lose interest!

Regarding the client version, you get that by running this (in thick mode only):

```
print(oracledb.clientversion())
```

**Reply by bnvader** [reply]
[db-trace.txt](https://github.com/oracle/python-oracledb/files/15354416/db-trace.txt)
Attached the trace for thin mode

**Reply by anthony-tuininga** [reply]
Thanks, can you also send the full traceback?

**Reply by bnvader** [reply]
How do I get full traceback?
Also client version is - 21, 6, 0, 0, 0

**Reply by anthony-tuininga** [reply]
If you do this in your Python program:

```python
1 / 0
```

You will get something like this:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ZeroDivisionError: division by zero
```

That is what I am looking for. If you do this, however:

```python
try:
    1 / 0
except Exception as e:
    print("ERROR:", e)
```

you will lose all of that information. If you can remove the try/except block and just let the exception get printed naturally, great. If not, you will have to look at the [traceback module](https://docs.python.org/3/library/traceback.html) to print it yourself.

**Reply by bnvader** [reply]
fetch_trasnactions
    transaction_xml_str = fetch_transaction_xml(cursor, resultobj.TRANS_ID)

  File ~\AppData\Local\anaconda3\Lib\site-packages\oracledb\dbobject.py:47 in __getattr__
    return self._impl.get_attr_value(attr_impl)

  File src\\oracledb\\impl/thin/dbobject.pyx:452 in oracledb.thin_impl.ThinDbObjectImpl.get_attr_value

  File src\\oracledb\\impl/thin/dbobject.pyx:162 in oracledb.thin_impl.ThinDbObjectImpl._ensure_unpacked

  File src\\oracledb\\impl/thin/dbobject.pyx:278 in oracledb.thin_impl.ThinDbObjectImpl._unpack_data

  File src\\oracledb\\impl/thin/dbobject.pyx:316 in oracledb.thin_impl.ThinDbObjectImpl._unpack_data_from_buf

  File src\\oracledb\\impl/thin/dbobject.pyx:346 in oracledb.thin_impl.ThinDbObjectImpl._unpack_value

  File src\\oracledb\\impl/base/buffer.pyx:723 in oracledb.base_impl.Buffer.read_str

UnicodeDecodeError: 'utf-8' codec can't decode byte 0x85 in position 0: invalid start byte

**Reply by anthony-tuininga** [reply]
Thank you. If you are able to tell me what the contents of the attribute are from running in thick mode that would be helpful in debugging this. I can probably figure it out from the packet data but if I know the "right" answer already, that will save me a lot of time!

**Reply by bnvader** [reply]
The first trans_id is 64262323 if that is what you meant.

**Reply by anthony-tuininga** [reply]
That helps, thanks. I now have a pretty good idea of what the problem is. Thin mode extracts all of the attributes in the object when the first attribute is accessed. Thick mode does not. It extracts just the attribute requested. You can confirm my suspicions by attempting to extract *all* of the attributes in thick mode -- and let me know which one fails!

**Reply by bnvader** [reply]
There is an XML type column in this. I m pretty sure that is the one causing the issue.
Even in thick mode, I was not able to retrieve that column and needed to write a separate select sql to retrieve it separately.
"select xmltype.getclobval(USER_DATA) from ebp.transaction where trans_id = " + str(trans_id)

**Reply by bnvader** [reply]
With thick client - 
print(resultobj.USER_DATA)
File ~\AppData\Local\anaconda3\Lib\site-packages\oracledb\dbobject.py:47 in __getattr__
return self._impl.get_attr_value(attr_impl)
File src\\oracledb\\impl/thick/dbobject.pyx:113 in oracledb.thick_impl.ThickDbObjectImpl.get_attr_value
File src\\oracledb\\impl/thick/utils.pyx:450 in oracledb.thick_impl._raise_from_odpi
File src\\oracledb\\impl/thick/utils.pyx:440 in oracledb.thick_impl._raise_from_info
DatabaseError: DPI-1008: data type 58 is not supported

**Reply by anthony-tuininga** [reply]
Thanks. I can replicate now and have created issue #336 to track this.

**Reply by bnvader** [reply]
Thank, appreciate it.

**Reply by anthony-tuininga** [reply]
A patch is available, @bnvader. See #336 for details.
